File: ios/Runner/Activity/SessionTiming.swift
import Foundation

/// Accumulated elapsed-recording-seconds for a session, shared by
/// `ActivityRecorderChannel` (pause/resume/stop/recover) and
/// `CoreLocationActivityRecorder` (GPS and timer scheduling) so the
/// two never drift apart.
///
/// Mirrors the Android `SessionTiming` object and the Dart geolocator
/// recorder: a paused session keeps its stored value; an active one adds the
/// current segment's running time measured from `resumedAt ?? startedAt`.
enum SessionTiming {
    static func elapsedSeconds(
        _ session: ActiveActivitySessionData,
        referenceMillis: Int64
    ) -> Int {
        if session.status != ActiveActivitySessionData.statusRecording {
            return session.elapsedDurationSeconds
        }
        guard let anchor = IsoTime.toEpochMillis(session.resumedAt ?? session.startedAt) else {
            return session.elapsedDurationSeconds
        }
        let segmentSeconds = Int((referenceMillis - anchor) / 1000)
        return session.elapsedDurationSeconds + max(0, segmentSeconds)
    }

    static func afterInterruption(
        _ session: ActiveActivitySessionData,
        lastPointMillis: Int64?,
        nowMillis: Int64
    ) -> ActiveActivitySessionData {
        guard session.status == ActiveActivitySessionData.statusRecording else {
            return session
        }
        let referenceMillis = lastPointMillis
            ?? IsoTime.toEpochMillis(session.resumedAt ?? session.startedAt)
            ?? nowMillis
        return session.copyWith(
            resumedAt: .some(IsoTime.format(Date(timeIntervalSince1970: Double(nowMillis) / 1000))),
            pausedAt: .some(nil),
            endedAt: .some(nil),
            elapsedDurationSeconds: elapsedSeconds(session, referenceMillis: referenceMillis)
        )
    }
}

File: android/app/src/main/kotlin/com/endurain/endurain/activity/SessionTiming.kt
package com.endurain.endurain.activity

import kotlin.math.max

/**
 * Accumulated elapsed-recording-seconds for a session, shared by
 * [ActivityRecorderChannel] (pause/resume/stop/recover) and
 * [ActivityRecorderService] (GPS and timer announcement scheduling) so the two
 * never drift apart.
 *
 * Mirrors the Dart geolocator recorder: a paused session keeps its stored
 * value; an active one adds the current segment's running time measured from
 * `resumedAt ?? startedAt`.
 */
object SessionTiming {
    fun elapsedSeconds(
        session: ActiveActivitySessionData,
        referenceMillis: Long,
    ): Int {
        if (session.status != ActiveActivitySessionData.STATUS_RECORDING) {
            return session.elapsedDurationSeconds
        }
        val anchor = IsoTime.toEpochMillis(session.resumedAt ?: session.startedAt)
            ?: return session.elapsedDurationSeconds
        val segmentSeconds = ((referenceMillis - anchor) / 1000L).toInt()
        return session.elapsedDurationSeconds + max(0, segmentSeconds)
    }

    fun afterInterruption(
        session: ActiveActivitySessionData,
        lastPointMillis: Long?,
        nowMillis: Long,
    ): ActiveActivitySessionData {
        if (session.status != ActiveActivitySessionData.STATUS_RECORDING) {
            return session
        }
        val referenceMillis = lastPointMillis
            ?: IsoTime.toEpochMillis(session.resumedAt ?: session.startedAt)
            ?: nowMillis
        return session.copy(
            resumedAt = IsoTime.format(java.util.Date(nowMillis)),
            pausedAt = null,
            endedAt = null,
            elapsedDurationSeconds = elapsedSeconds(session, referenceMillis),
        )
    }
}
